packages/@glimmer/core/src/template.ts

```diff
@@ -1,92 +1,25 @@
-
-export interface TemplateMeta {
-  scope: () => Dict<unknown>;
-}
-export type TemplateScope = () => Dict<unknown>;
-
-const TEMPLATE_MAP = new WeakMap<object, SerializedTemplateWithLazyBlock>();
-const getPrototypeOf = Object.getPrototypeOf;
-
-export const SCOPE_MAP = new WeakMap<SerializedTemplateWithLazyBlock, TemplateScope>();
-
-export interface CustomSerializedTemplate {
-  id?: string | null;
-  block: SerializedTemplateBlockJSON;
-  meta: TemplateMeta;
-}
+import { setComponentTemplate as vmSetComponentTemplate } from '@glimmer/manager';
+import { SerializedTemplateWithLazyBlock, Dict } from '@glimmer/interfaces';
+import { templateFactory } from '@glimmer/opcode-compiler';
 
 // This is provided by the `babel-plugin-strict-template-precompile` plugin
 export let createTemplate: (
   scopeOrTemplate: Dict<unknown> | string,
   template?: string
-) => CustomSerializedTemplate;
+) => SerializedTemplateWithLazyBlock;
 
 if (DEBUG) {
-  createTemplate = (): CustomSerializedTemplate => {
+  createTemplate = (): SerializedTemplateWithLazyBlock => {
     throw new Error(
       'createTemplate() is meant to be preprocessed with a babel plugin, @glimmer/babel-plugin-strict-template-precompile. If you are seeing this error message, it means that you do not have this babel plugin installed, or it is not enabled correctly'
     );
   };
 }
 
-const builtInHelpers = {
-  if: true,
-  each: true,
-};
-
-export function setComponentTemplate<T extends object>(
-  template: CustomSerializedTemplate,
-  ComponentClass: T
-): T {
-  const {
-    id,
-    block,
-    meta: { scope },
-  } = template;
-
-  if (DEBUG) {
-    const evaluatedScope = scope();
-    const parsed = JSON.parse(block);
-
-    if (parsed.upvars) {
-      for (const upvar of parsed.upvars) {
-        if (!(upvar in builtInHelpers) && evaluatedScope[upvar] === undefined) {
-          throw new Error(
-            `Cannot find identifier \`${upvar}\` in scope. It was used in a template, but not imported into the template scope or defined as a local variable. If you meant to access a property, you must add \`this\` to it: \`{{this.${upvar}}}\``
-          );
-        }
-      }
-    }
-  }
-
-  const newTemplate = { id, block, moduleName: '(unknown module)' };
-
-  SCOPE_MAP.set(newTemplate, scope);
-
-  TEMPLATE_MAP.set(ComponentClass, newTemplate);
-  return ComponentClass;
-}
-
-export function getComponentTemplate<T extends object>(
-  ComponentClass: T
-): SerializedTemplateWithLazyBlock | undefined {
-  let pointer = ComponentClass;
-
-  while (pointer !== undefined && pointer !== null) {
-    const manager = TEMPLATE_MAP.get(pointer);
-
-    if (manager !== undefined) {
-      return manager;
-    }
-
-    pointer = getPrototypeOf(pointer);
-  }
-
-  return undefined;
+export function setComponentTemplate(
+  template: SerializedTemplateWithLazyBlock,
+  ComponentClass: object
+): object {
+  return vmSetComponentTemplate(templateFactory(template), ComponentClass);
```

Comment:
we should likely update babel transforms to make this unnecessary, the VM version matches the public API version in Ember, so it'd be good to make sure they stay aligned. Again, fine to do in a followup, and if the motivation is to save bytes we could expose a different compile-target function that does this.

Reply:
https://github.com/glimmerjs/glimmer.js/issues/326